server/lib/payments.js

```diff
@@ -154,49 +154,65 @@ export async function createRefundTransaction(transaction, refundedPaymentProces
-   * we're just setting them to positive by adding a - sign in front
-   * of it. */
-  userLedgerRefund.hostFeeInHostCurrency = -userLedgerRefund.hostFeeInHostCurrency;
-  userLedgerRefund.platformFeeInHostCurrency = -userLedgerRefund.platformFeeInHostCurrency;
-  userLedgerRefund.paymentProcessorFeeInHostCurrency = -userLedgerRefund.paymentProcessorFeeInHostCurrency;
-
-  /* If the payment processor doesn't refund the fee, the equivalent
-   * of the fee will be transferred from the host to the user so the
-   * user can get the full refund. */
-  if (refundedPaymentProcessorFee === 0) {
-    userLedgerRefund.hostFeeInHostCurrency += userLedgerRefund.paymentProcessorFeeInHostCurrency;
-    userLedgerRefund.paymentProcessorFeeInHostCurrency = 0;
-  }
+  const buildRefund = t => {
+    const refund = pick(t, [
+      'currency',
+      'FromCollectiveId',
+      'CollectiveId',
+      'HostCollectiveId',
+      'PaymentMethodId',
+      'OrderId',
+      'hostCurrencyFxRate',
+      'hostCurrency',
+      'hostFeeInHostCurrency',
+      'platformFeeInHostCurrency',
+      'paymentProcessorFeeInHostCurrency',
+      'data.isFeesOnTop',
+    ]);
+    refund.CreatedByUserId = user.id;
+    refund.description = `Refund of "${t.description}"`;
+    refund.data = data;
+
+    /* The refund operation moves back fees to the user's ledger so the
+     * fees there should be positive. Since they're usually in negative,
+     * we're just setting them to positive by adding a - sign in front
+     * of it. */
+    refund.hostFeeInHostCurrency = -refund.hostFeeInHostCurrency;
+    refund.platformFeeInHostCurrency = -refund.platformFeeInHostCurrency;
+    refund.paymentProcessorFeeInHostCurrency = -refund.paymentProcessorFeeInHostCurrency;
+
+    /* If the payment processor doesn't refund the fee, the equivalent
+     * of the fee will be transferred from the host to the user so the
+     * user can get the full refund. */
+    if (refundedPaymentProcessorFee === 0) {
+      refund.hostFeeInHostCurrency += refund.paymentProcessorFeeInHostCurrency;
+      refund.paymentProcessorFeeInHostCurrency = 0;
+    }
+
+    /* Amount fields. Must be calculated after tweaking all the fees */
+    refund.amount = -t.amount;
+    refund.amountInHostCurrency = -t.amountInHostCurrency;
+    refund.netAmountInCollectiveCurrency = -libtransactions.netAmount(t);
+    return refund;
+  };
 
-  /* Amount fields. Must be calculated after tweaking all the fees */
-  userLedgerRefund.amount = -collectiveLedger.amount;
-  userLedgerRefund.amountInHostCurrency = -collectiveLedger.amountInHostCurrency;
-  userLedgerRefund.netAmountInCollectiveCurrency = -libtransactions.netAmount(collectiveLedger);
+  const creditTransactionRefund = buildRefund(creditTransaction);
+
+  if (transaction.data?.isFeesOnTop) {
+    const feeOnTopTransaction = await models.Transaction.findOne({
+      where: { ...FEES_ON_TOP_TRANSACTION_PROPERTIES, type: 'CREDIT' },
```

Comment:
This `FEES_ON_TOP_TRANSACTION_PROPERTIES` export is pretty handy. You may be interested in Sequelize [scopes](https://sequelize.org/master/manual/scopes.html) for similar needs in the future.